docs/spec/enums.rst

```diff
@@ -0,0 +1,363 @@
+        pass
+
+    class Color7(CustomEnum1):  # Supported
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+    class CustomEnumType(EnumType):
+        pass
+
+    class CustomEnum2(metaclass=CustomEnumType):
+        pass
+
+    class Color8(CustomEnum2):  # Supported
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+
+Enum Behaviors
+--------------
+
+Enum classes are iterable and indexable, and they can be called with a value
+to look up the enum member with that value. Type checkers should support these
+behaviors::
+
+    class Color(Enum):
+        RED = 1
+        GREEN = 2
+        BLUE = 3
+
+    for color in Color:
+        reveal_type(color)  # Revealed type is 'Color'
+
+    reveal_type(Color["RED"])  # Revealed type is 'Literal[Color.RED]' (or 'Color')
+    reveal_type(Color(3))  # Revealed type is 'Literal[Color.BLUE]' (or 'Color')
+
+Unlike most Python classes, Calling an enum class does not invoke its constructor.
+Instead, the call performs a value-based lookup of an enum member.
+
+An Enum class with one or more defined members cannot be subclassed. They are
+implicitly "final". Type checkers should enforce this::
+
+    class EnumWithNoMembers(Enum):
+        pass
+
+    class Shape(EnumWithNoMembers):  # OK (because no members are defined)
+        SQUARE = 1
+        CIRCLE = 2
+
+    class ExtendedShape(Shape):  # Type checker error: Shape is implicitly final
+        TRIANGLE = 3
+
+
+Defining Members
+----------------
+
+When using the "class syntax", enum classes can define both members and
+other (non-member) attributes. The ``EnumType`` metaclass applies a set
+of rules to distinguish between members and non-members. Type checkers
+should honor the most common of these rules. The lesser-used rules are
+optional. Some of these rules may be impossible to evaluate and enforce
+statically in cases where dynamic values are used.
+
+* If an attribute is defined in the class body with a type annotation but
+  with no assigned value, a type checker should assume this is a non-member
+  attribute::
+
+    class Pet(Enum):
+        genus: str  # Non-member attribute
+        species: str  # Non-member attribute
+
+        CAT = 1  # Member attribute
+        DOG = 2  # Member attribute
+
+  Within a type stub, members can be defined using the actual runtime values,
+  or a placeholder of ``...`` can be used::
+
+    class Pet(Enum):
+        genus: str  # Non-member attribute
+        species: str  # Non-member attribute
+
+        CAT = ...  # Member attribute
+        DOG = ...  # Member attribute
+
+* Members defined within an enum class should not include explicit type
+  annotations. Type checkers should infer a literal type for all members.
+  A type checker should report an error if a type annotation is used
+  for an enum member because this type will be incorrect and misleading
+  to readers of the code::
+
+    class Pet(Enum):
+        CAT = 1  # OK
+        DOG: int = 2  # Type checker error
```

Comment:
Yes, this should be a type checker error in that case as well. Any annotation should be flagged as an error. 

There's a fundamental problem with annotating symbols corresponding to enum member types: their types change dynamically due to the magic of the EnumMeta class. Initially, when executing the class body, these symbols are assigned a value based on the assignment expression (e.g. `int` or `Literal[2]` in the example above). After the class body is executed, these symbols are overwritten by the EnumMeta class with a different value that is an instance of the enum class. Because of this dynamic runtime behavior, there's no annotation that accurately describes the type of an enum member.